src/uct/ib/mlx5/ib_mlx5.c

```diff
@@ -348,8 +401,106 @@ ucs_status_t uct_ib_mlx5_get_compact_av(uct_ib_iface_t *iface, int *compact_av)
 }
 #endif
 
-void uct_ib_mlx5_check_completion(uct_ib_iface_t *iface, uct_ib_mlx5_cq_t *cq,
-                                  struct mlx5_cqe64 *cqe)
+static void
+uct_ib_mlx5_iface_cqe_unzip_fill_unique(struct mlx5_cqe64 *cqe,
+                                        uct_ib_mlx5_mini_cqe8_t *mini_cqe,
+                                        uct_ib_mlx5_cq_unzip_t *cq_unzip)
+{
+    const uint32_t net_qpn_mask = htonl(UCS_MASK(UCT_IB_QPN_ORDER));
+    uint16_t new_wqe_opcode;
+
+    /* TODO: Think about the checksum field unzipping
+     * The checksum field does not copied from src dst
+     * because of absence of the mlx5_cqe64::check_sum field on some systems.
+     */
+    cqe->byte_cnt = mini_cqe->byte_cnt;
+    ucs_assert(!(cqe->op_own & UCT_IB_MLX5_CQE_OP_OWN_ERR_MASK));
+    if ((cqe->op_own >> 4) == MLX5_CQE_REQ) {
+        cqe->wqe_counter  = mini_cqe->wqe_counter;
+        cqe->sop_drop_qpn = (cqe->sop_drop_qpn & net_qpn_mask) |
+                            htonl(mini_cqe->s_wqe_opcode << UCT_IB_QPN_ORDER);
+    } else if (cq_unzip->block_size != cq_unzip->left) {
+        /* The title_cqe opcode shouldn't be updated */
+        new_wqe_opcode = ntohs(cqe->wqe_counter) + 1;
+        cqe->wqe_counter = htons(new_wqe_opcode);
+    }
+}
+
+struct mlx5_cqe64 *uct_ib_mlx5_iface_cqe_unzip_continue(uct_ib_mlx5_cq_t *cq)
+{
+    uct_ib_mlx5_cq_unzip_t *cq_unzip  = &cq->cq_unzip;
+    uint8_t mini_cqe_idx              = (cq_unzip->block_size -
+                                         cq_unzip->left) %
+                                        UCT_IB_MLX5_MINICQE_ARR_MAX_SIZE;
+    struct mlx5_cqe64 *title_cqe      = &cq_unzip->title;
+    uct_ib_mlx5_mini_cqe8_t *mini_cqe = &cq_unzip->mini_arr[mini_cqe_idx];
+    struct mlx5_cqe64 *cqe;
+
+    uct_ib_mlx5_iface_cqe_unzip_fill_unique(title_cqe, mini_cqe, cq_unzip);
+
+    cq_unzip->left--;
+    cq->cq_ci++;
+
+    if (cq_unzip->left != 0) {
+        if (mini_cqe_idx == (UCT_IB_MLX5_MINICQE_ARR_MAX_SIZE - 1)) {
+            /* Get the next mini_cqe array */
+            cqe = uct_ib_mlx5_get_cqe(cq, cq->cq_ci);
+            memcpy(&cq_unzip->mini_arr, cqe, sizeof(*cqe));
```

Comment:
do we know haw many cqes in this block so can copy less?
for example copy min(cq_unzip->remaining*sizeof(..), sizeof(*cqe))
BTW, seems that according to PRM, there can be up to 7 mini-cqes, and not 8